dhcpdump.8
.\" Automatically generated by Pod::Man 4.11 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "DHCPDUMP 8"
.TH DHCPDUMP 8 "23 June 2008" "perl v5.30.0" "User Contributed Software"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
dhcpdump \- DHCP packet dumper
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBdhcpdump\fR [\fB\-h\fR \fIregular-expression\fR] <\fB\-i\fR \fIinterface\fR | \fB\-r\fR \fIpcap_file\fR>
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This command parses the output of tcpdump to display the dhcp-packets for
easier checking and debugging.
.SH "USAGE"
.IX Header "USAGE"
for live capture:
.PP
dhcpdump\ \-i\ /dev/fxp0
.PP
or for \s-1PCAP\s0 file:
.PP
dhcpdump\ \-r\ pcap-file-name
.PP
If you want to filter a specific Client Hardware Address (\s-1CHADDR\s0), then
you can specifiy it as a regular expressions:
.PP
dhcpdump\ \-i\ /dev/fxp0\ \-h\ ^00:c0:4f
.PP
This will display only the packets with Client Hardware Addresses
which start with 00:c0:4f.
.SH "OUTPUT"
.IX Header "OUTPUT"
.Vb 10
\&  TIME: 15:45:02.084272
\&    IP: 0.0.0.0.68 (0:c0:4f:82:ac:7f) > 255.255.255.255.67 (ff:ff:ff:ff:ff:ff)
\&    OP: 1 (BOOTPREQUEST)
\& HTYPE: 1 (Ethernet)
\&  HLEN: 6
\&  HOPS: 0
\&   XID: 28f61b03
\&  SECS: 0
\& FLAGS: 0
\&CIADDR: 0.0.0.0
\&YIADDR: 0.0.0.0
\&SIADDR: 0.0.0.0
\&GIADDR: 0.0.0.0
\&CHADDR: 00:c0:4f:82:ac:7f:00:00:00:00:00:00:00:00:00:00
\& SNAME: .
\& FNAME: .
\&OPTION:  53 (  1) DHCP message type         3 (DHCPREQUEST)
\&OPTION:  54 (  4) Server identifier         130.139.64.101
\&OPTION:  50 (  4) Request IP address        130.139.64.143
\&OPTION:  55 (  7) Parameter Request List      1 (Subnet mask)
\&                                              3 (Routers)
\&                                             58 (T1)
\&                                             59 (T2)
.Ve
.PP
At the option field, the first field is the value of the option,
the second one (between brackets) is the length of the option-datafield,
the third field is the name of the option, the fourth field is the
data of the option.
.SH "RETURN VALUES"
.IX Header "RETURN VALUES"
Always 0.
.SH "NOTES"
.IX Header "NOTES"
Privileged access is often needed for accessing the interface.
.SH "BUGS"
.IX Header "BUGS"
Not all the parameter options are printed verbose, because of lack of
documentation. Not all the options are tested, because of lack of
clients/servers with these options. If you have a dump of one of
them, please send them to me and I'll incorperate them.
.SH "THANKS TO"
.IX Header "THANKS TO"
Ralph Droms and Ted Lemons \*(L"The \s-1DHCP\s0 Handbook\*(R", \s-1ISBN 1\-57870\-137\-6.\s0
.PP
Peter Apian-Bennewitz <apian@ise.fhg.de> for his Client Hardware
Address filtering
.SH "AUTHOR"
.IX Header "AUTHOR"
Edwin Groothuis, edwin@mavetju.org (http://www.mavetju.org)
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBdhcpd\fR\|(8), \fBtcpdump\fR\|(1), \s-1RFC2132\s0


dhcpdump.c
// DHCPDUMP
//
// Usage: tcpdump -s 1518 -lenx port bootps or port bootpc | dhcpdump
//
// note 1: how does this work for FDDI / PPP links?
// note 2: what is this number 14?
//
// $Id: dhcpdump.c,v 1.12 2004/10/31 11:22:58 mavetju Exp $
//

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <net/ethernet.h>
#include <netinet/in_systm.h>
#include <netinet/in.h>
#include <netinet/ip.h>
#include <netinet/udp.h>
#include <netinet/ether.h>
#include <arpa/inet.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <time.h>
#include <pcap.h>
#include <err.h>
#include <regex.h>
#include "dhcp_options.h"

#ifndef HAVE_STRSEP
#include "strsep.c"
#endif

#define SPERW	(7 * 24 * 3600)
#define SPERD	(24 * 3600)
#define SPERH	(3600)
#define SPERM	(60)

#define LARGESTRING 1024

// header variables
char	timestamp[40];			// timestamp on header
char	mac_origin[40];			// mac address of origin
char	mac_destination[40];		// mac address of destination
char	ip_origin[40];			// ip address of origin
char	ip_destination[40];		// ip address of destination
int	max_data_len;			// maximum size of a packet

int	tcpdump_style = -1;
char	errbuf[PCAP_ERRBUF_SIZE];
char	*hmask = NULL;
regex_t	preg;

int	check_ch(u_char *data, int data_len);
int	readheader(u_char *buf);
int	readdata(u_char *buf, u_char *data, int *data_len);
int	printdata(u_char *data, int data_len);
void	pcap_callback(u_char *user, const struct pcap_pkthdr *h,
	    const u_char *sp);

void	printIPaddress(u_char *data);
void	printIPaddressAddress(u_char *data);
void	printIPaddressMask(u_char *data);
void	print8bits(u_char *data);
void	print16bits(u_char *data);
void	print32bits(u_char *data);
void	printTime8(u_char *data);
void	printTime32(u_char *data);
void	printReqParmList(u_char *data, int len);
void	printHexColon(u_char *data, int len);
void	printHex(u_char *data, int len);
void	printHexString(u_char *data, int len);
void	printGUID(u_char *data);

void usage(char *me) {
	fprintf(stderr,"Usage: %s <-i interface | -r tcpdump_file> [-h macaddress]\n",me);
	exit(1);
}

int main(int argc, char **argv) {
	int	i;
	pcap_t *cap;
	struct bpf_program fp;
	char	*interface = NULL;
	char    *offline_file = NULL;

	for (i = 1; i < argc; i++) {
		if (argv[i] == NULL || argv[i][0] != '-') break;
		switch (argv[i][1]) {
		case 'h':
			hmask = argv[++i];
			break;
		case 'i':
			interface = argv[++i];
			break;
		case 'r':
			offline_file = argv[++i];
			break;
		default:
			fprintf(stderr, "%c: uknown option\n",argv[i][1]);
			usage(argv[0]);
		}
	}

	if (interface == NULL && offline_file == NULL) usage(argv[0]);
	if (interface != NULL && offline_file != NULL) usage(argv[0]);

	if (hmask)
		regcomp(&preg, hmask, REG_EXTENDED | REG_ICASE | REG_NOSUB);

	if(interface!=NULL) {
		if ((cap = pcap_open_live(interface, 1500, 1, 100, errbuf)) == NULL)
			errx(1, "pcap_open_live(): %s", errbuf);
	} else {
		if ((cap = pcap_open_offline(offline_file,errbuf)) == NULL)
			errx(1, "unable to open file %s", errbuf);
	}
	if (pcap_compile(cap, &fp, "udp and (port bootpc or port bootps)", 0, 0) < 0)
		errx(1,"pcap_compile: %s", pcap_geterr(cap));
	if (pcap_setfilter(cap, &fp) < 0)
		errx(1,"pcap_setfilter: %s", pcap_geterr(cap));
	if (pcap_loop(cap, 0, pcap_callback, NULL) < 0)
		errx(1,"pcap_loop(%s): %s", interface, pcap_geterr(cap));

	return 0;
}

void pcap_callback(u_char *user, const struct pcap_pkthdr *h, const u_char *sp) {
	struct ether_header *eh;
	struct ip *ip;
	struct udphdr *udp;
	int offset = 0;

	if (h->caplen < ETHER_HDR_LEN) {
		printf("Ignored too short ethernet packet: %d bytes\n",
		    h->caplen);
		return;
	}

	eh = (struct ether_header *)(sp + offset);
	offset += ETHER_HDR_LEN;

	// Check for IPv4 packets
	if (eh->ether_type != 8) { 
		printf("Ignored non IPv4 packet: %d\n", eh->ether_type);
		return;
	}

	// Check for length
	if (h->caplen < offset + sizeof(struct ip)) {
		printf("Ignored too short IPv4 packet: %d bytes\n", h->caplen);
		return;
	}

	ip = (struct ip *)(sp + offset);
	offset += sizeof(struct ip);

	udp = (struct udphdr *)(sp + offset);
	offset += sizeof(struct udphdr);

	{
		struct timeval tp;
		gettimeofday(&tp, NULL);
		strftime(timestamp, sizeof(timestamp), "%Y-%m-%d %H:%M:%S.",
		    localtime(&(tp.tv_sec)));
		sprintf(timestamp + strlen(timestamp), "%03ld",
		    tp.tv_usec / 1000);
	}

	strcpy(mac_origin, ether_ntoa((struct ether_addr *)eh->ether_shost));
	strcpy(mac_destination,
	    ether_ntoa((struct ether_addr *)eh->ether_dhost));

	strcpy(ip_origin, (char *)inet_ntoa(ip->ip_src));
	strcpy(ip_destination, (char *)inet_ntoa(ip->ip_dst));

#ifdef __linux__
	if (hmask && check_ch((u_char *)(sp + offset), ntohs(udp->len)))
		return;

	printdata((u_char *)(sp + offset), ntohs(udp->len));
#else
	if (hmask && check_ch((u_char *)(sp + offset), ntohs(udp->uh_ulen)))
		return;

	printdata((u_char *)(sp + offset), ntohs(udp->uh_ulen));
#endif

}

// check for matching CHADDR (Peter Apian-Bennewitz <apian@ise.fhg.de>)
int check_ch(u_char *data, int data_len) {
	char ch_ip[50];

	if (data_len<43) return(0);
	sprintf(ch_ip,
	    "%02x:%02x:%02x:%02x:%02x:%02x:%02x:%02x:"
	    "%02x:%02x:%02x:%02x:%02x:%02x:%02x:%02x\n",
	    data[28], data[29], data[30], data[31],
	    data[32], data[33], data[34], data[35],
	    data[36], data[37], data[38], data[39],
	    data[40], data[41], data[42], data[43]);
	return (regexec(&preg, ch_ip, 0, NULL, 0));
}


// print the data as an IP address
void printIPaddress(u_char *data) {
	printf("%d.%d.%d.%d", data[0], data[1], data[2], data[3]);
}

// print the data as an IP address and an IP address
void printIPaddressAddress(u_char *data) {
	printf("%d.%d.%d.%d %d.%d.%d.%d",
	    data[0], data[1], data[2], data[3],
	    data[4], data[5], data[6], data[7]);
}

// print the data as an IP address and mask
void printIPaddressMask(u_char *data) {
	printf("%d.%d.%d.%d/%d.%d.%d.%d",
	    data[0], data[1], data[2], data[3],
	    data[4], data[5], data[6], data[7]);
}

// prints a value of 8 bits (1 byte)
void print8bits(u_char *data) {
	printf("%d", data[0]);
}

// prints a value of 16 bits (2 bytes)
void print16bits(u_char *data) {
	printf("%d", (data[0] << 8) + data[1]);
}

// prints a value of 32 bits (4 bytes)
void print32bits(u_char *data) {
	printf("%d",
	    (data[0] << 24) + (data[1] << 16) + (data[2] << 8) + data[3]);
}

// print the data as a 8bits time-value
void printTime8(u_char *data) {
	int t = data[0];
	printf("%d (", t);
	if (t > SPERW) { printf("%dw", t / (SPERW)); t %= SPERW; }
	if (t > SPERD) { printf("%dd", t / (SPERD)); t %= SPERD; }
	if (t > SPERH) { printf("%dh", t / (SPERH)); t %= SPERH; }
	if (t > SPERM) { printf("%dm", t / (SPERM)); t %= SPERM; }
	if (t > 0) printf("%ds", t);
	printf(")");
}

// print the data as a 32bits time-value
void printTime32(u_char *data) {
	int t = (data[0] << 24) + (data[1] << 16) + (data[2] <<8 ) + data[3];
	printf("%d (", t);
	if (t > SPERW) { printf("%dw", t / (SPERW)); t %= SPERW; }
	if (t > SPERD) { printf("%dd", t / (SPERD)); t %= SPERD; }
	if (t > SPERH) { printf("%dh", t / (SPERH)); t %= SPERH; }
	if (t > SPERM) { printf("%dm", t / (SPERM)); t %= SPERM; }
	if (t > 0) printf("%ds", t);
	printf(")");
}

// print the data as a hex-list, with the translation into ascii behind it
void printHexString(u_char *data, int len) {
	int i, j, k;

	for (i = 0; i <= len / 8; i++) {
		for (j = 0; j < 8; j++) {
			if (i * 8 + j >= len) break;
			printf("%02x", data[i * 8 + j]);
		}
		for (k = j; k < 8; k++)
			printf("  ");
		printf(" ");
		for (j = 0; j < 8; j++) {
			char c = data[i * 8 + j];
			if (i * 8 + j >= len) break;
			printf("%c", isprint(c) ? c : '.');
		}
		if (i * 8 + j < len) printf("\n\t\t\t\t\t      ");
	}
}

// print the data as a hex-list, without the translation into ascii behind it
void printHex(u_char *data, int len) {
	int i, j;

	for (i = 0; i <= len / 8; i++) {
		for (j = 0; j < 8; j++) {
			if (i * 8 + j >= len) break;
			printf("%02x", data[i * 8 + j]);
		}
		if (i * 8 + j < len) printf("\n\t\t\t\t\t    ");
	}
}

// print the data as a hex-list seperated by colons
void printHexColon(u_char *data, int len) {
	int i;

	for (i = 0; i < len; i++) {
		if (i != 0 ) printf(":");
		printf("%02x", data[i]);
	}
}

// print the list of requested parameters
void printReqParmList(u_char *data, int len) {
	int i;

	for (i = 0; i < len; i++) {
		printf("%3d (%s)\n", data[i], dhcp_options[data[i]]);
		printf("\t\t\t\t\t      ");
	}
}

void printGUID(u_char *data) {
	printf("%02X%02X%02X%02X",data[0],data[1],data[2],data[3]);
	data += 4;
	printf("-%02X%02X-%02X%02X-%02X%02X",data[0],data[1],data[2],data[3],data[4],data[5]);
	data += 6;
	printf("%02X%02X%02X%02X%02X%02X",data[0],data[1],data[2],data[3],data[4],data[5]);
}

// print the header and the options.
int printdata(u_char *data, int data_len) {
	int	j, i;
	char	buf[LARGESTRING];
	uint16_t *arch_type;

	if (data_len == 0)
		return 0;

	printf(  "  TIME: %s\n", timestamp);
	printf(  "    IP: %s (%s) > %s (%s)\n",
	    ip_origin, mac_origin, ip_destination, mac_destination);
	printf(  "    OP: %d (%s)\n", data[0], operands[data[0]]);
	printf(  " HTYPE: %d (%s)\n", data[1], htypes[data[1]]);
	printf(  "  HLEN: %d\n", data[2]);
	printf(  "  HOPS: %d\n", data[3]);
	printf(  "   XID: %02x%02x%02x%02x\n",
	    data[4], data[5], data[6], data[7]);
	printf(  "  SECS: "); print16bits(data + 8);
	printf("\n FLAGS: %x\n", 255 * data[10] + data[11]);

	printf(  "CIADDR: "); printIPaddress(data + 12);
	printf("\nYIADDR: "); printIPaddress(data + 16);
	printf("\nSIADDR: "); printIPaddress(data + 20);
	printf("\nGIADDR: "); printIPaddress(data + 24);
	printf("\nCHADDR: "); printHexColon(data+28, 16);
	printf("\n SNAME: %s.\n", data + 44);
	printf(  " FNAME: %s.\n", data + 108);

	j = 236;
	j += 4;	/* cookie */
	while (j < data_len && data[j] != 255) {
		printf("OPTION: %3d (%3d) %-28s", data[j], data[j + 1],
		    dhcp_options[data[j]]);

	switch (data[j]) {
	default:
		printHexString(data + j + 2, data[j + 1]);
		break;

	case 0:		// pad
		break;

	case  1:	// Subnetmask
	case  3:	// Routers
	case 16:	// Swap server
	case 28:	// Broadcast address
	case 32:	// Router solicitation
	case 50:	// Requested IP address
	case 54:	// Server identifier
		printIPaddress(data + j + 2);
		break;

	case 12:	// Hostname
	case 14:	// Merit dump file
	case 15:	// Domain name
	case 17:	// Root Path
	case 18:	// Extensions path
	case 40:	// NIS domain
	case 56:	// Message
	case 62:	// Netware/IP domain name
	case 64:	// NIS+ domain
	case 66:	// TFTP server name
	case 67:	// bootfile name
	case 60:	// Domain name
	case 86:	// NDS Tree name
	case 87:	// NDS context
		strncpy(buf, (char *)&data[j + 2], data[j + 1]);
		buf[data[j + 1]] = 0;
		printf("%s", buf);
		break;

	case  4:	// Time servers
	case  5:	// Name servers
	case  6:	// DNS server
	case  7:	// Log server
	case  8:	// Cookie server
	case  9:	// LPR server
	case 10:	// Impress server
	case 11:	// Resource location server
	case 41:	// NIS servers
	case 42:	// NTP servers
	case 44:	// NetBIOS name server
	case 45:	// NetBIOS datagram distribution server
	case 48:	// X Window System font server
	case 49:	// X Window System display server
	case 65:	// NIS+ servers
	case 68:	// Mobile IP home agent
	case 69:	// SMTP server
	case 70:	// POP3 server
	case 71:	// NNTP server
	case 72:	// WWW server
	case 73:	// Finger server
	case 74:	// IRC server
	case 75:	// StreetTalk server
	case 76:	// StreetTalk directory assistance server
	case 85:	// NDS server
		for (i = 0; i < data[j + 1] / 4; i++) {
			if (i != 0) printf(",");
			printIPaddress(data + j + 2 + i * 4);
		}
		break;

	case 21:	// Policy filter
		for (i = 0; i < data[j + 1] / 8; i++) {
			if (i != 0) printf(",");
			printIPaddressMask(data + j + 2 + i * 8);
		}
		break;

	case 33:	// Static route
		for (i = 0; i < data[j + 1] / 8; i++) {
			if (i != 0) printf(",");
			printIPaddressAddress(data + j + 2 + i * 8);
		}
		break;

	case 25:	// Path MTU plateau table
		for (i = 0; i < data[j + 1] / 2; i++) {
			if (i != 0) printf(",");
			print16bits(data + j + 2 + i * 2);
		}
		break;

	case 13:	// bootfile size
	case 22:	// Maximum datagram reassembly size
	case 26:	// Interface MTU
	case 57:	// Maximum DHCP message size
		print16bits(data + j + 2);
		break;

	case 19:	// IP forwarding enabled/disable
	case 20:	// Non-local source routing
	case 27:	// All subnets local
	case 29:	// Perform mask discovery
	case 30:	// Mask supplier
	case 31:	// Perform router discovery
	case 34:	// Trailer encapsulation
	case 39:	// TCP keepalive garbage
		printf("%d (%s)", data[j + 2], enabledisable[data[j + 2]]);
		break;

	case 23:	// Default IP TTL
		printTime8(data + j + 2);
		break;

	case 37:	// TCP default TTL
		print8bits(data + j + 2);
		break;

	case 43:	// Vendor specific info
	case 47:	// NetBIOS scope (no idea how it looks like)
		printHexString(data + j + 2, data[j + 1]);
		break;

	case 46:	// NetBIOS over TCP/IP node type
		printf("%d (%s)",
		    data[j + 2], netbios_node_type[data[j + 2]]);
		break;
	    
	case  2:	// Time offset
	case 24:	// Path MTU aging timeout
	case 35:	// ARP cache timeout
	case 38:	// TCP keepalive interval
	case 51:	// IP address leasetime
	case 58:	// T1
	case 59:	// T2
		printTime32(data + j + 2);
		break;

	case 36:	// Ethernet encapsulation
		printf("%d (%s)",
		    data[j + 2],
		    data[j +2 ] > sizeof(ethernet_encapsulation) ?
			"*wrong value*" :
			ethernet_encapsulation[data[j + 2]]);
		break;

	case 52:	// Option overload
		printf("%d (%s)",
		    data[j + 2],
		    data[j + 2] > sizeof(option_overload) ?
			"*wrong value*" :
			option_overload[data[j + 2]]);
		break;

	case 53:	// DHCP message type
		printf("%d (%s)",
		    data[j + 2],
		    data[j + 2] > sizeof(dhcp_message_types) ?
			"*wrong value*" :
			dhcp_message_types[data[j + 2]]);
		break;

	case 55:	// Parameter Request List
		printReqParmList(data + j + 2, data[j + 1]);
		break;

	case 63:	// Netware/IP domain information
		printHex(data + j + 2, data[j + 1]);
		break;

	case 61:	// Client identifier
		printHexColon(data + j + 2, data[j + 1]);
		break;

	case 81:	// Client FQDN
		print8bits(data + j + 2);
		printf("-");
		print8bits(data + j + 3);
		printf("-");
		print8bits(data + j + 4);
		printf(" ");
		strncpy(buf, (char *)&data[j + 5], data[j + 1] - 3);
		buf[data[j + 1] - 3]=0;
		printf("%s", buf);
		break;

	case 82:	// Relay Agent Information
		printf("\n");
		for (i = j + 2; i < j + data[j + 1]; ) {
			if (i != j+2) {
				printf("\n");
			}
			printf("%-17s %-13s ", " ",
			    data[i] > sizeof(relayagent_suboptions) ?
			    "*wrong value*" :
			    relayagent_suboptions[data[i]]);
			if (i + data[i + 1] > j + data[j + 1]) {
				printf("*MALFORMED -- TOO LARGE*\n");
				break;
			}
			printHexColon(data + i + 2, data[i + 1]);
			i += data[i + 1] + 2;
		}
		break;
	case 97:	// Client Machine Identifier
		if(data[j+1]==17 && data[j+2]==0) {
			printGUID(data+j+3);
		} else {
			printHexString(data + j + 2, data[j + 1]);
		}
		break;
	case 94:	//Client Network Interface Identifier
		{
			char _type=data[j+2];
			char _M=data[j+3];
			char _m=data[j+4];
			if(_type==1) {
				printf("UNDI %d.%d",_M,_m);
			} else {
				printf("UNKNOW TYPE");
			}
			printf(" (");
			printHexColon(data + j + 2,3);
			printf(")");
		}
		break;
	case 93:	// Client System Architecture Type
		for(i = j + 2; i <= j + data[j + 1];i+=2) {
			arch_type=(uint16_t *)(data+i);
			// https://www.ietf.org/assignments/dhcpv6-parameters/dhcpv6-parameters.xml
			// Belowe definitions are for DHCPv6 option 61 but they are compatible with DHCPv4 option
			// see https://tools.ietf.org/html/rfc5970#section-3.3
			// EDK 2 definies few of them
			// I only saw  arch type 0x10 on x64 UEFI servers with DHCPv4
			switch(ntohs(*arch_type)) {
				case 0x00:
					printf("x86 BIOS");
					break;
				case 0x01:
					printf("NEC/PC98");		// DEPRECATED
					break;
				case 0x02:
					printf("Itanium");
					break;
				case 0x03:
					printf("DEC Alpha");		// DEPRECATED
					break;
				case 0x04:
					printf("Arc x86");		// DEPRECATED
					break;
				case 0x05:
					printf("Intel Lean Client");	// DEPRECATED
					break;
				case 0x06:
					printf("x86 UEFI");			
					break;
				case 0x07:
					printf("x64 UEFI");
					break;
				case 0x08:
					printf("EFI Xscale");		// DEPRECATED
					break;
				case 0x09:
					printf("EBC");
					break;
				case 0x0A:
					printf("ARM 32-bit UEFI");
					break;
				case 0x0B:
					printf("ARM 64-bit UEFI");
					break;
				case 0x0C:
					printf("PowerPC Open Firmware");
					break;
				case 0x0D:
					printf("PowerPC ePAPR");
					break;
				case 0x0E:
					printf("POWER OPAL v3");
					break;
				case 0x0F:
					printf("x86 uefi boot from http");
					break;
				case 0x10:
					printf("x64 uefi boot from http");
					break;
				case 0x11:
					printf("ebc boot from http");
					break;
				case 0x12:
					printf("arm uefi 32 boot from http");
					break;
				case 0x13:
					printf("arm uefi 64 boot from http");
					break;
				case 0x14:
					printf("pc/at bios boot from http");
					break;
				case 0x15:
					printf("arm 32 uboot");
					break;
				case 0x16:
					printf("arm 64 uboot");
					break;
				case 0x17:
					printf("arm uboot 32 boot from http");
					break;
				case 0x18:
					printf("arm uboot 64 boot from http");
					break;
				case 0x19:
					printf("RISC-V 32-bit UEFI");
					break;
				case 0x1A:
					printf("RISC-V 32-bit UEFI boot from http");
					break;
				case 0x1B:
					printf("RISC-V 64-bit UEFI");
					break;
				case 0x1C:
					printf("RISC-V 64-bit UEFI boot from http");
					break;
				case 0x1D:
					printf("RISC-V 128-bit UEFI");
					break;
				case 0x1E:
					printf("RISC-V 128-bit UEFI boot from http");
					break;
				case 0x1F:
					printf("s390 Basic");
					break;
				case 0x20:
					printf("s390 Extended");
					break;
				case 0x21:
					printf("MIPS 32-bit UEFI");
					break;
				case 0x22:
					printf("MIPS 64-bit UEFI");
					break;
				case 0x23:
					printf("Sunway 32-bit UEFI");
					break;
				case 0x24:
					printf("Sunway 64-bit UEFI");
					break;
				default:
					printf("UNASSIGNED");
			}
			printf(" (");
			printHexColon(data + i, 2);
			printf(")");
			if(i + 2 <= j + data[j+1]) {
				printf(", ");
			}
		}
		break;
	}
	printf("\n");

	/*
	// This might go wrong if a mallformed packet is received.
	// Maybe from a bogus server which is instructed to reply
	// with invalid data and thus causing an exploit.
	// My head hurts... but I think it's solved by the checking
	// for j<data_len at the begin of the while-loop.
	*/
	if (data[j]==0)		// padding
		j++;
	else
		j+=data[j + 1] + 2;

	}

	puts("---------------------------------------------------------------------------\n");
	fflush(stdout);

	return 0;
}
